tests/test_s3urlcache.py

```diff
@@ -0,0 +1,189 @@
+        bad_urls = ['', '//', 'http://?']
+        self.urls_to_cleanup.extend(bad_urls)
+
+        for url in bad_urls:
+            with self.subTest(bad_url=url):
+                with self.assertRaises(requests.RequestException):
+                    self.cache.resolve(url)
+
+    def test_url_SizeLimitError(self):
+        #  Exception returned when url content size is greater than max_size.
+        url = f"{HTTPInfo.url}/{KB}"
+        self.urls_to_cleanup.append(url)
+        self.cache.max_size = 1
+        with self.assertRaises(SizeLimitError) as ex:
+            self.cache.resolve(url)
+        self.assertEqual(ex.exception.args[0], f"{url} not cached. The URL's contents have exceeded "
+                                               f"{self.cache.max_size} bytes.")
+
+    def test_content_size_0(self):
+        url = f"{HTTPInfo.url}/0"
+        self.urls_to_cleanup.append(url)
+        url_content = self.cache.resolve(url)
+        self.assertEqual(len(url_content), 0)
+
+    def test_chunked_content(self):
+        size = KB * 10
+        url = f"{HTTPInfo.url}/{size}"
+        self.urls_to_cleanup.append(url)
+        self.cache.max_size = MB
+        self.cache.chunk_size = KB
+        url_content = self.cache.resolve(url)
+        original_data = RANDOMDATA[:size]
+        self.assertEqual(url_content, original_data)
+
+    def test_stored_url_metadata(self):
+        url = f"{HTTPInfo.url}/{KB}"
+        self.urls_to_cleanup.append(url)
+        url_key = S3UrlCache._create_key(url)
+        self.cache.resolve(url)
+        with self.subTest("check dss_cached_url"):
+            metadata = self.cache.reverse_key_lookup(url_key)
+            self.assertEqual(metadata, {'dss_cached_url': url})
+
+        with self.subTest("check content_type"):
+            contentType = self.blobstore.get_content_type(self.test_bucket, url_key)
+            self.assertEqual(contentType, "application/octet-stream")
+
+    def test_invalidate(self):
+        url = f"{HTTPInfo.url}/{KB}"
+        self.urls_to_cleanup.append(url)
+        url_key = S3UrlCache._create_key(url)
+
+        with self.assertLogs(logger, "INFO") as log_monitor:
+            # Verify the URL is cached
+            self.cache.resolve(url)
+            # Remove the URL from cache
+            self.cache.invalidate(url)
+            # Check that it was added to cache again.
+            self.cache.resolve(url)
+        self.assertTrue(log_monitor.output[0].endswith(f"{url} not found in cache. Added it to "
+                                                       f"{self.test_bucket} with key {url_key}."))
+        self.assertTrue(log_monitor.output[1].endswith(f"{url} removed from cache in {self.test_bucket}."))
+        self.assertTrue(log_monitor.output[2].endswith(f"{url} not found in cache. Added it to "
+                                                       f"{self.test_bucket} with key {url_key}."))
+
+    def delete_cached_urls(self, urls):
```

Comment:
This can be protected.

Reply:
do you really need to worry about protecting functions with testcases? Fixed.

Reply:
It makes it more obvious that this is a utility method and not a test case. We do have a few abstract test cases (test_indexer) where this is naming convention is also useful.